it/src/main/java/com/google/cloud/teleport/it/datastream/DefaultDatastreamResourceManager.java

```diff
@@ -0,0 +1,362 @@
+                                    .setSourceConfig(sourceConfig)
+                                    .setDisplayName(streamId)
+                                    .setDestinationConfig(destinationConfig)
+                                    .setBackfillAll(Stream.BackfillAllStrategy.getDefaultInstance())
+                                    .build())
+                            .build();
+
+            Stream reference = datastreamClient.createStreamAsync(request).get();
+            createdStreamIds.add(streamId);
+
+            return reference;
+        } catch (ExecutionException | InterruptedException e) {
+            throw new DatastreamResourceManagerException("Failed to create stream. ", e);
+        }
+    }
+
+    public Stream updateStreamState(String streamId, Stream.State state) {
+
+        LOG.info("Updating {}'s state to {} in project {}.", streamId, state.name(), projectId);
+
+        try {
+            Stream.Builder streamBuilder = Stream.newBuilder()
+                    .setName(StreamName.format(projectId, location, streamId))
+                    .setState(state);
+
+            FieldMask.Builder fieldMaskBuilder = FieldMask.newBuilder().addPaths(FIELD_STATE);
+
+            UpdateStreamRequest request = UpdateStreamRequest.newBuilder()
+                    .setStream(streamBuilder)
+                    .setUpdateMask(fieldMaskBuilder)
+                    .build();
+
+            return datastreamClient.updateStreamAsync(request).get();
+        } catch (InterruptedException | ExecutionException e) {
+            throw new DatastreamResourceManagerException("Failed to update stream. ", e);
+        }
+    }
+
+    @Override
+    public Stream startStream(String streamId) {
+        LOG.info("Starting Stream {} in project {}.", streamId, projectId);
+        return updateStreamState(streamId, Stream.State.RUNNING);
+    }
+
+    @Override
+    public Stream pauseStream(String streamId) {
+        LOG.info("Pausing Stream {} in project {}.", streamId, projectId);
+        return updateStreamState(streamId, Stream.State.PAUSED);
+    }
+
+    @Override
+    public void cleanupAll() {
+        try {
+            for (String stream : createdStreamIds) {
+                datastreamClient.deleteStreamAsync(DeleteStreamRequest
+                        .newBuilder()
+                        .setName(StreamName.format(projectId, location, stream))
+                        .build()).get();
+            }
+
+            for (String connectionProfile : createdConnectionProfileIds) {
+                datastreamClient.deleteConnectionProfileAsync(DeleteConnectionProfileRequest
+                        .newBuilder().setName(ConnectionProfileName.format(projectId, location, connectionProfile))
+                        .build()).get();
+            }
+        } catch (InterruptedException | ExecutionException e) {
+            throw new DatastreamResourceManagerException("Failed to delete datastream resources. ", e);
+        }
```

Comment:
I would put these delete loops in separate try-catch blocks to try and close as many resources as possible even if there is an error. Look at how the `MongDBResourceManager's` `cleanupAll()` works.
https://github.com/GoogleCloudPlatform/DataflowTemplates/blob/f00c2f2ccbebe56fbd9569c00aaac4736a519af1/it/mongodb/src/main/java/com/google/cloud/teleport/it/mongodb/MongoDBResourceManager.java#L242-L274

Reply:
There also should be a `datastreamClient.close()` or similar function that closes the client's connection to the resource that needs to be invoked here. I've seen issues with new clients being unable to connect to resources in subsequent tests due to unclosed connection streams, so I want to avoid that here.